frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,814 @@
+		self.solution
+			.score
+			.cmp(&other.solution.score)
+			.then_with(|| self.solution.cmp(&other.solution))
+			.then_with(|| self.deposit.cmp(&other.deposit))
+			.then_with(|| self.reward.cmp(&other.reward))
+			.then_with(|| self.who.cmp(&other.who))
+	}
+}
+
+impl<AccountId, Balance, CompactSolution> PartialOrd
+	for SignedSubmission<AccountId, Balance, CompactSolution>
+where
+	AccountId: Ord,
+	Balance: Ord + HasCompact,
+	CompactSolution: Ord,
+	RawSolution<CompactSolution>: Ord,
+{
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+pub type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::PositiveImbalance;
+pub type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+pub type SignedSubmissionOf<T> =
+	SignedSubmission<<T as frame_system::Config>::AccountId, BalanceOf<T>, CompactOf<T>>;
+
+pub type SubmissionIndicesOf<T> =
+	BoundedBTreeMap<ElectionScore, u32, <T as Config>::SignedMaxSubmissions>;
+
+/// Mask type which pretends to be a set of `SignedSubmissionOf<T>`, while in fact delegating to the
+/// actual implementations in `SignedSubmissionIndices<T>`, `SignedSubmissionsMap<T>`, and
+/// `SignedSubmissionNextIndex<T>`.
+#[cfg_attr(feature = "std", derive(DebugNoBound))]
+pub struct SignedSubmissions<T: Config> {
+	indices: SubmissionIndicesOf<T>,
+	next_idx: u32,
+	insertion_overlay: BTreeMap<u32, SignedSubmissionOf<T>>,
+	deletion_overlay: BTreeSet<u32>,
+}
+
+impl<T: Config> SignedSubmissions<T> {
+	/// Get the signed submissions from storage.
+	pub fn get() -> Self {
+		SignedSubmissions {
+			indices: SignedSubmissionIndices::<T>::get(),
+			next_idx: SignedSubmissionNextIndex::<T>::get(),
+			insertion_overlay: BTreeMap::new(),
+			deletion_overlay: BTreeSet::new(),
+		}
+	}
+
+	/// Put the signed submissions back into storage.
+	pub fn put(self) {
+		SignedSubmissionIndices::<T>::put(self.indices);
+		SignedSubmissionNextIndex::<T>::put(self.next_idx);
+		for key in self.deletion_overlay {
+			SignedSubmissionsMap::<T>::remove(key);
+		}
+		for (key, value) in self.insertion_overlay {
+			SignedSubmissionsMap::<T>::insert(key, value);
+		}
+	}
+
+	/// Get the submission at a particular index.
+	fn get_submission(&self, idx: u32) -> Option<SignedSubmissionOf<T>> {
+		self.insertion_overlay
+			.get(&idx)
+			.cloned()
+			.or_else(|| SignedSubmissionsMap::<T>::try_get(idx).ok())
+	}
+
+	/// Take the submission at a particular index.
+	fn take_submission(&mut self, idx: u32) -> Option<SignedSubmissionOf<T>> {
+		self.insertion_overlay.remove(&idx).or_else(|| {
+			if self.deletion_overlay.contains(&idx) {
+				None
+			} else {
+				self.deletion_overlay.insert(idx);
+				SignedSubmissionsMap::<T>::try_get(idx).ok()
+			}
+		})
```

Comment:
This function doesn't update indices, I know it is not public, and its current usage is correct. but it is still accessible from this module, So the doc should warn about this specific behavior.

Reply:
Updated the doc in https://github.com/paritytech/substrate/pull/7910/commits/0b0abb8c7ab839493147ed822b67d08da56eaba6; is that what you had in mind?

As for removing it from indices in this function, I see what you're getting at, but IMO it's easier to reason about what precisely is happening when there are fewer side effects.

Reply:
the way I see it is that there is less "side effects" when all function goes from a valid state of the variable to a valid state of the variable.

about the doc I don't understand, this function does examine the insertion overlay, what it does not is updating the indices field of the struct.

Reply:
The doc should be:
"take submission update the fields `insertion_overlay` and `deletion_overlay` but doesn't update the field `indices`."

Reply:
Ok; 955f5bcce.

> there is less "side effects" when all function goes from a valid state of the variable to a valid state of the variable.

That's a very good argument, and I tried it your way. The problem is when we go to use it later:

https://github.com/paritytech/substrate/blob/955f5bccea40e74ca63ca4d0c55c6c1214fdedbe/frame/election-provider-multi-phase/src/signed.rs#L246-L253

We need to update `self.indices` by removing the weakest element, at which point we know we have enough space to insert the new one. However, we want the output of `self.take_submission` to be the ultimate output of the function. Doing things your way looks like this:

```rust
let weakest_submission = self.take_submission(weakest_idx);
// `self.take_submission` also removes the weakest idx from `self.indices`
self.indices 
 	.try_insert(score, insert_idx) 
 	.expect("just removed an item, we must be under capacity; qed"); 
weakest_submission
```

IMO this makes it harder to prove the correctness of the `.expect`, and overall seems less elegant.

Reply:
I think a minor flaw with `SignedSubmissions` struct wrapper is that it is trying to look like a concrete general purpose API, while it is still not there. I agree that the current call-paths seem correct, but the naming are such that they give you the impression that a function `x` is designed to be used in all normal circumstances. 

I think a simple and reasonable way with minimal code-change is to rename the function in such a way that these API quirks is reflected, next to the doc update.

I had the same comment about `decode_length` _**not**_ having the same name as our storage `decode_length` to avoid a similar problem.

Reply:
As for this particular case, I think what can help you is a `swap` function that simultaneously updates all relevant storage items, and preferably removing `take`.

Reply:
https://github.com/paritytech/substrate/pull/7910/commits/368e97bd2a3aec3ff99bea19d04ff3dc7d8ea3af implements the suggested change, removing `take` and adding `swap`. I agree that this is a more general solution. 

I don't really understand the problem with `decode_len`, though; as far as I can tell, it has precisely the same semantics: it reads storage exactly enough to decode the number of items contained.